realm-jni/src/io_realm_internal_TableQuery.cpp

```diff
@@ -859,21 +861,219 @@ JNIEXPORT jlong JNICALL Java_io_realm_internal_TableQuery_nativeFind(
     return -1;
 }
 
+std::unique_ptr<Query> getHandoverQuery (jlong bgSharedGroupPtr, jlong replicationPtr, jlong queryPtr) {
+    SharedGroup::Handover <Query> *handoverQueryPtr = reinterpret_cast<SharedGroup::Handover <Query> * > (queryPtr);
+    std::unique_ptr <SharedGroup::Handover<Query>> handoverQuery(handoverQueryPtr);
+
+    SG(bgSharedGroupPtr)->end_read();
+
+    SharedGroup::VersionID currentVersion = SG(bgSharedGroupPtr)->get_version_of_current_transaction();
+    bool isDifferentVersions = (currentVersion != handoverQuery->version);
+    if (isDifferentVersions) {
+        SG(bgSharedGroupPtr)->begin_read(handoverQuery->version);
+    } else {
+        SG(bgSharedGroupPtr)->begin_read();
+    }
+
+    std::unique_ptr <Query> query = SG(bgSharedGroupPtr)->import_from_handover(std::move(handoverQuery));
+    if (isDifferentVersions) {
+        LangBindHelper::advance_read(*SG(bgSharedGroupPtr), *CH(replicationPtr));
+    }
+
+    return query;
+}
+
+JNIEXPORT jlong JNICALL Java_io_realm_internal_TableQuery_nativeFindWithHandover(
+    JNIEnv* env, jobject, jlong bgSharedGroupPtr, jlong replicationPtr, jlong queryPtr, jlong fromTableRow)
+{
+    TR_ENTER()
+    try {
+        std::unique_ptr <Query> query = getHandoverQuery (bgSharedGroupPtr, replicationPtr, queryPtr);
+        Table* table = query->get_table().get();
+
+        if (!QUERY_VALID(env, query.get())) {
+            return 0;
+        }
+
+        // It's valid to go 1 past the end index
+        if ((fromTableRow < 0) || (S(fromTableRow) > table->size())) {
+            // below check will fail with appropriate exception
+            (void) ROW_INDEX_VALID(env, table, fromTableRow);
+            return 0;
+        }
+
+        size_t r = query->find(S(fromTableRow));
+        if (r == not_found) {
+            return 0;
```

Comment:
add a comment here https://github.com/realm/realm-java/issues/1489
need to generalise this everywhere